### PR TITLE
Sebastien servo control

### PR DESCRIPTION
**Pull request by sssmc**
Added more features to servo class

### CODE REVIEW

File: INMOOV-Head-Control/StandardServo.cpp
Comment (Gordon-928) [inline INMOOV-Head-Control/StandardServo.cpp:10]: - What is the intention of using two constructor in the cpp file?
- Add default arguments to the constructor StandardServo::StandardServo() (just a reminder, ignore if you plan to add it later)

File: INMOOV-Head-Control/StandardServo.h
Comment (gordonso928) [inline INMOOV-Head-Control/StandardServo.h]: How this value is obtained? Is it default value?

File: INMOOV-Head-Control/StandardServo.cpp
Comment (gordonso928) [inline INMOOV-Head-Control/StandardServo.cpp]: 1. What is the intention of using two constructor in the cpp file?

2. Add default arguments to the constructor StandardServo::StandardServo() (just a reminder, ignore if you plan to add it later)

3. Add two arguments m_minPWM & m_maxPWM (pending names), as different servo will have different min & max PWM value

File: INMOOV-Head-Control/StandardServo.cpp
Comment (gordonso928) [inline INMOOV-Head-Control/StandardServo.cpp]: use m_minPWM & m_maxPWM in the map() function

File: INMOOV-Head-Control/StandardServo.cpp
Comment (gordonso928) [inline INMOOV-Head-Control/StandardServo.cpp]: use m_minPWM for calculations

**Review by gordonso928 — CHANGES_REQUESTED** [review]
The code generally looks very good, a few changes can make the code more readable:

- [x] Adding comments to variables & functions, simple comments would be enough (I believe you are planning to do it afterwards)
- [x] Adding explanation of how certain value are calculated (i.e. m_PWMPerDegree)
- [x] Some of the functions should be placed on top (e.g. moveToTargetMicroseconds() is called by a lot of functions). Let me know the intention if we need to keep it this way.

File: INMOOV-Head-Control/StandardServo.h
Comment (sssmc) [reply INMOOV-Head-Control/StandardServo.h]: Yes this would be the default value. This is the full range of the servo in microseconds divided by the the full range of the servo in degrees. This is mostly a temporary solution, still trying to figure out a better way of doing this.

**Comment by sssmc** [comment]
Not 100% sure what you mean by this: "Some of the functions should be placed on top " Are you referring the the order of the functions in the .cpp/.h files?

**Comment by sssmc** [comment]
I ran out of time for comments yesterday. I sent the PR so that we can start reviewing right a way, I will get started on them later today or tomorrow.

**Comment by gordonso928** [comment]
RE Order: Yes, that is correct 
RE Comments: Yeah this is what I think what happened when I make the comment. We can add it later as long as other knows how to use the functions

**Comment by sssmc** [comment]
> Yes, that is correct

Ok, so the functions should go in order of use? Or order of most used?

File: INMOOV-Head-Control/StandardServo.cpp
Comment (sssmc) [reply INMOOV-Head-Control/StandardServo.cpp]: Will do

File: INMOOV-Head-Control/StandardServo.cpp
Comment (sssmc) [reply INMOOV-Head-Control/StandardServo.cpp]: Will do

**Comment by gordonso928** [comment]
Should the function be declared before calling it?

File: INMOOV-Head-Control/StandardServo.cpp
Comment (sssmc) [reply INMOOV-Head-Control/StandardServo.cpp]: This is my first time using c++, so I trying to figure the conventions. I have added defaults for all values in the header file. The first constructor just takes the pin number and leaves all other values default. The second constructor allows the user to specify all values.

Should the defaults be specified in the constructor(s)?  

Is there a way to have optional parameters in c++?

File: INMOOV-Head-Control/StandardServo.h
Comment (gordonso928) [reply INMOOV-Head-Control/StandardServo.h]: This would be determined from the specification of the servo it would be:
PWMPerDegree = (maxPWM - minPWM)/servoDegreeRange

Comment (sssmc) [reply INMOOV-Head-Control/StandardServo.h]: I think maybe this should be changed to some sort of gear ratio, then the class could convert into a PWM per degree. This would be needed since some of the servos are geared and I want the user to be able to specify the amount of degrees that the robot would move and not have to worry about how much the servos need to move to achieve this movement. 

The min/max values should be also be specified in robot rotation not servo rotation

**Comment by sssmc** [comment]
> Should the function be declared before calling it?

I think it works either way, but I agree it makes more sense that way.

File: INMOOV-Head-Control/StandardServo.h
Comment (gordonso928) [reply INMOOV-Head-Control/StandardServo.h]: Ok that sounds good, taking account in the gear ration would be a good call. This would be equivalent of multiplying the control gain of the servo and gears. This maybe a better approach as the user can set up the object by using the specifications of the servo and any gear combination. For example:

_**Servo Specs**_
image

Servo Running Degree: 270 degree
PWM Range: 500 - 2500 microseconds
--> Calculate the Servo Gain

_**Gear**_
Gear Ratio: 2:1
--> Gain = 2

This is just an example, double check how to set it up correctly. If you are still stuck I can help you with that. For now I think we can just put place holder.

Comment (sssmc) [reply INMOOV-Head-Control/StandardServo.h]: Ok, il give it a go, il let you know if I am having any troubles

File: INMOOV-Head-Control/StandardServo.cpp
Comment (gordonso928) [reply INMOOV-Head-Control/StandardServo.cpp]: I would just use default arguments, if the user did no input them, they will be set to the default value.

Comment (sssmc) [reply INMOOV-Head-Control/StandardServo.cpp]: Ok, I will look into this

**Comment by sssmc** [comment]
Me and @Kayleb814 refactored the entire movement system to use a PWM min/max value and a gear ratio. Added a good amount of comments as well explaining things.

Still don't think I am setting up the constructor/default values correctly, you might have to explain how you want it more specially.

All functions seems to be working as intended with our limited testing.

**Comment by gordonso928** [comment]
> Me and @Kayleb814 refactored the entire movement system to use a PWM min/max value and a gear ratio. Added a good amount of comments as well explaining things.
> 
> 
> 
> Still don't think I am setting up the constructor/default values correctly, you might have to explain how you want it more specially.
> 
> 
> 
> All functions seems to be working as intended with our limited testing.   

That looks very good! If you can use the Servo.h with the current code, forget about the constructor thingy. However, it is most likely you will get a compile error if you are using platformio. I will take a closer look at the code and merger the PR tonight.

**Comment by sssmc** [comment]
> > Me and @Kayleb814 refactored the entire movement system to use a PWM min/max value and a gear ratio. Added a good amount of comments as well explaining things.
> > Still don't think I am setting up the constructor/default values correctly, you might have to explain how you want it more specially.
> > All functions seems to be working as intended with our limited testing.
> 
> That looks very good! If you can use the Servo.h with the current code, forget about the constructor thingy. However, it is most likely you will get a compile error if you are using platformio. I will take a closer look at the code and merger the PR tonight.

We have been testing in Arduino IDE, pure c++ is pickier about some things so you most likely correct.

One more thing I think we should decide on before merging: Currently the user specifies a min and max rotation of the part controlled by the servo. But when a angle is specified it is in relation to the entire movement range of the part, the code will just constrain the move to the min/max range specified. Is this the behavior that we want or do we want the input angle for a move to be relative to the min/max range(ie. 0 degrees=min rotation, (maxRotation - minRotation)= max rotation)?

**Comment by gordonso928** [comment]
> > > Me and @Kayleb814 refactored the entire movement system to use a PWM min/max value and a gear ratio. Added a good amount of comments as well explaining things.
> 
> > > Still don't think I am setting up the constructor/default values correctly, you might have to explain how you want it more specially.
> 
> > > All functions seems to be working as intended with our limited testing.
> 
> > 
> 
> > That looks very good! If you can use the Servo.h with the current code, forget about the constructor thingy. However, it is most likely you will get a compile error if you are using platformio. I will take a closer look at the code and merger the PR tonight.
> 
> 
> 
> We have been testing in Arduino IDE, pure c++ is pickier about some things so you most likely correct.
> 
> 
> 
> One more thing I think we should decide on before merging: Currently the user specifies a min and max rotation of the part controlled by the servo. But when a angle is specified it is in relation to the entire movement range of the part, the code will just constrain the move to the min/max range specified. Is this the behavior that we want or do we want the input angle for a move to be relative to the min/max range(ie. 0 degrees=min rotation, (maxRotation - minRotation)= max rotation)?

I believe we want to constrain the movement range of the servo based on the control range. For example, if a part have a range of 0-90 degree, while the servo only need to move 0-45 degree to achieve that movement. We would want to constrain the servo range to 0-45 degree, even it can move 0-180 degree. Also the "zero" of the movement is not fixed and we can define our own, use the most convenient/proper zero will be fine.

For the constructor, forget it as ArduinoIDE works fine. Keep going and the code will eventually come together.

**Comment by sssmc** [comment]
> > > > Me and @Kayleb814 refactored the entire movement system to use a PWM min/max value and a gear ratio. Added a good amount of comments as well explaining things.
> > 
> > 
> > > > Still don't think I am setting up the constructor/default values correctly, you might have to explain how you want it more specially.
> > 
> > 
> > > > All functions seems to be working as intended with our limited testing.
> > 
> > 
> > > 
> > 
> > 
> > > That looks very good! If you can use the Servo.h with the current code, forget about the constructor thingy. However, it is most likely you will get a compile error if you are using platformio. I will take a closer look at the code and merger the PR tonight.
> > 
> > 
> > We have been testing in Arduino IDE, pure c++ is pickier about some things so you most likely correct.
> > One more thing I think we should decide on before merging: Currently the user specifies a min and max rotation of the part controlled by the servo. But when a angle is specified it is in relation to the entire movement range of the part, the code will just constrain the move to the min/max range specified. Is this the behavior that we want or do we want the input angle for a move to be relative to the min/max range(ie. 0 degrees=min rotation, (maxRotation - minRotation)= max rotation)?
> 
> I believe we want to constrain the movement range of the servo based on the control range. For example, if a part have a range of 0-90 degree, while the servo only need to move 0-45 degree to achieve that movement. We would want to constrain the servo range to 0-45 degree, even it can move 0-180 degree. Also the "zero" of the movement is not fixed and we can define our own, use the most convenient/proper zero will be fine.
> 
> For the constructor, forget it as ArduinoIDE works fine. Keep going and the code will eventually come together.

Ok, I will change to work that way, it will be in a separate commit, so we can always change it back if needed.

**Comment by sssmc** [comment]
Standard Servo class should be feature complete and working based on our limited testing. Let me know if there is any other changes you would like us to make. We should also do some more in depth testing at a later date to make sure there isn't any edge cases that will cause problems.

**Review by gordonso928 — APPROVED** [review]
Look good! I will merge this PR and we can test that later.

**Pull request merged**
